src/api.js

```diff
@@ -0,0 +1,1003 @@
+            var i;
+            var result;
+            var value_ptr;
+            var value_type;
+            args = [];
+            function data_func_blob(ptr) {
+                var blob_arg;
+                var blob_ptr;
+                var j;
+                var size;
+                size = sqlite3_value_bytes(ptr);
+                blob_ptr = sqlite3_value_blob(ptr);
+                blob_arg = new Uint8Array(size);
+                j = 0;
+                while (j < size) {
+                    blob_arg[j] = HEAP8[blob_ptr + j];
+                    j += 1;
+                }
+                return blob_arg;
+            }
+            function data_func_null() {
+                return null;
+            }
+            i = 0;
+            while (i < argc) {
+                value_ptr = getValue(argv + (4 * i), "i32");
+                value_type = sqlite3_value_type(value_ptr);
+                switch (value_type) {
+                case 1:
+                    data_func = sqlite3_value_double;
+                    break;
+                case 2:
+                    data_func = sqlite3_value_double;
+                    break;
+                case 3:
+                    data_func = sqlite3_value_text;
+                    break;
+                case 4:
+                    data_func = data_func_blob;
+                    break;
+                default:
+                    data_func = data_func_null;
+                }
+                arg = data_func(value_ptr);
+                args.push(arg);
+                i += 1;
+            }
+            try {
+                result = func.apply(null, args);
+            } catch (error1) {
+                error = error1;
+                sqlite3_result_error(cx, error, -1);
+                return;
+            }
+            switch (typeof result) {
+            case "boolean":
+                sqlite3_result_int(cx, result ? 1 : 0);
+                break;
+            case "number":
+                sqlite3_result_double(cx, result);
+                break;
+            case "string":
+                sqlite3_result_text(cx, result, -1, -1);
+                break;
+            case "object":
+                if (result === null) {
+                    sqlite3_result_null(cx);
+                } else if (!isNullOrUndefined(result.length)) {
+                    blobptr = allocate(result, "i8", ALLOC_NORMAL);
+                    sqlite3_result_blob(cx, blobptr, result.length, -1);
+                    _free(blobptr);
+                } else {
+                    sqlite3_result_error(cx, (
+                        "Wrong API use : tried to return a value "
+                            + "of an unknown type (" + result + ")."
+                    ), -1);
+                }
+                break;
+            default:
+                sqlite3_result_null(cx);
+            }
+        }
+        if (Object.prototype.hasOwnProperty.call(this.functions, name)) {
```

Comment:
`this.function` is a normal-object.  its simpler to call hasOwnProperty normally:
```diff
@@ 615
         this.functions = {};
@@ 978
-if (Object.prototype.hasOwnProperty.call(this.functions, name)) {
+if (this.functions.hasOwnProperty(name)) {
```

Reply:
The keys of this.function are controlled by the user.. The code would then break if someone added a function called 'hasOwnProperty'.